pkgs/stdenv/generic/make-derivation.nix

```
@@ -163,6 +163,31 @@ let
 
 , env ? { }
 
+, src ? null
+
+, srcWorkDir ? null
+
+, sourceRoot ?
+    if srcWorkDir == null then
+      null
+    else if ! builtins.isPath srcWorkDir then
+      throw "stdenv.mkDerivation: The `srcWorkDir` attribute is expected to be a path, but it's a ${builtins.typeOf srcWorkDir} instead."
+    else if ! builtins.pathExists srcWorkDir then
+      throw "stdenv.mkDerivation: The `srcWorkDir` attribute \"${toString srcWorkDir}\" does not exist."
+    else if lib.filesystem.pathType srcWorkDir != "directory" then
+      throw "stdenv.mkDerivation: The `srcWorkDir` attribute \"${toString srcWorkDir}\" is not a directory"
+    # All of the following conditions are always fulfilled if `src = lib.fileset.toSource { ... }`
+    else if ! src._isLibCleanSourceWith or false then
+      throw "stdenv.mkDerivation: If the `srcWorkDir` attribute is set, `src` must be a source-like value produced by the functions in `lib.sources`."
+    else if ! builtins.isPath src.origSrc then
+      throw "stdenv.mkDerivation: If the `srcWorkDir` attribute is set, `src` must be produced from a local path."
+    else if ! lib.path.hasPrefix src.origSrc srcWorkDir then
+      throw "stdenv.mkDerivation: The `srcWorkDir` \"${toString srcWorkDir}\" needs to be under the local path that produced the source \"${src.origSrc}\"."
```

```suggestion
      throw "stdenv.mkDerivation: If the `srcWorkDir` attribute is set, `src` must be produced from a local directory."
    else if ! lib.path.hasPrefix src.origSrc srcWorkDir then
      throw "stdenv.mkDerivation: The `srcWorkDir` \"${toString srcWorkDir}\" must be within the local directory that is the source of \"${src.origSrc}\"."
```

Something being "under a path" doesn't really make sense to me.